src/LanguageServer/Protocol/Features/Options/SolutionAnalyzerConfigOptionsUpdater.WorkspaceUpdater.cs

```diff
@@ -0,0 +1,234 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Options;
+
+internal sealed partial class SolutionAnalyzerConfigOptionsUpdater
+{
+    /// <summary>
+    /// We need to update the current solution snapshot based on two sources of events:
+    /// 1) global options service (global option changed)
+    /// 2) the workspace (project added)
+    /// 
+    /// To keep the set of analyzer fallback options on the solution snapshot consistent and not miss any option 
+    /// we serialize the updates using a lock.
+    /// </summary>
+    private sealed class WorkspaceUpdater(Workspace workspace, EditorConfigOptionsEnumerator optionsEnumerator, IGlobalOptionService globalOptions)
+    {
+        // Set of languages (keys of the dictionary) in the current snapshot that have their fallback options initialized from global options.
+        // The values are ignored.
+        private ImmutableDictionary<string, int> _initializedLanguages = ImmutableDictionary<string, int>.Empty;
+
+        // Ensures we only process one update at a time:
+        private readonly object _updateLock = new();
+
+        public void WorkspaceChanged(object? sender, WorkspaceChangeEventArgs args)
+        {
+            try
+            {
+                switch (args.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionChanged:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                    case WorkspaceChangeKind.ProjectAdded:
+                    case WorkspaceChangeKind.ProjectChanged:
+                    case WorkspaceChangeKind.ProjectReloaded:
+                    case WorkspaceChangeKind.ProjectRemoved:
+                        break;
+
+                    default:
+                        return;
+                }
+
+                if (args.OldSolution.SolutionState.ProjectCountByLanguage.KeysEqual(args.NewSolution.SolutionState.ProjectCountByLanguage))
```

Comment:
minor nit, but do we need to listen to project-changed/reloaded?  might want to doc if that's the case (for example, if that can change the project language count).